python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -235,14 +245,205 @@ private Node update(Node node) {
+ * ```
+ * where `[` and `]` denotes synthesisezed nodes, so `[*t]` is the synthesized tuple argument
+ * `TPosOverflowNode` and `[**d]` is the synthesized dictionary argument `TKwOverflowNode`.
+ * There will be a store step from `2` to `[*t]` at pos `0` and one from `3` to `[**d]` at key
+ * `a`.
+ *
+ * For the call
+ * ```python
+ * f(0, **{"y": 1, "a": 3})
+ * ```
+ * no tuple arguemnt is synthesized. It is modelled as
+ * ```python
+ * f(0, [y=1], [**d])
+ * ```
+ * where `[y=1]` is the synthesised unpacked argument `TKwUnpacked` (with `name` = `y`). There is
+ * a read step from `**{"y": 1, "a": 3}` to `[y=1]` at key `y` to get the value passed to the parameter
+ * `y`. There is a dataflow step from `**{"y": 1, "a": 3}` to `[**d]` to transfer the content and
+ * a clearing of content at key `y` for node `[**d]`, since that value has been unpacked.
+ */
+module ArgumentPassing {
+  /**
+   * Holds if `call` is a call to `callable`.
+   * Used to limit the size of predicates.
+   */
+  predicate connects(CallNode call, CallableValue callable) {
+    exists(DataFlowCall c |
+      call = c.getNode() and
+      callable = c.getCallable().getCallableValue()
+    )
+  }
+
+  /**
+   * Gets the `n`th parameter of `callable`.
+   * If the callable has a starred parameter, say `*tuple`, that is matched with `n=-1`.
+   * If the callable has a doubly starred parameter, say `**dict`, that is matched with `n=-2`.
+   * Note that, unlike other languages, we do _not_ use -1 for the position of `self` in Python,
+   * as it is an explicit parameter at position 0.
+   */
+  NameNode getParameter(CallableValue callable, int n) {
+    // positional parameter
+    result = callable.getParameter(n)
+    or
+    // starred parameter, `*tuple`
+    exists(Function f |
+      f = callable.getScope() and
+      n = -1 and
+      result = f.getVararg().getAFlowNode()
+    )
+    or
+    // doubly starred parameter, `**dict`
+    exists(Function f |
+      f = callable.getScope() and
+      n = -2 and
+      result = f.getKwarg().getAFlowNode()
+    )
+  }
+
+  /**
+   * Gets the argument to `call` that is passed to the `n`th parameter of `callable`.
+   * If it is a positional argument, it must appear at position `argNr`.
+   * `argNr` will differ from `n` for method- or class calls, where the first parameter
+   * is `self` and the first positional argument is passed to the second positional parameter.
+   */
+  Node getArg(CallNode call, int argNr, CallableValue callable, int n) {
+    connects(call, callable) and
+    n - argNr in [0, 1] and // constrain for now to limit the size of the predicate; we only use it to insert one argument (self).
+    (
+      // positional argument
+      result = TCfgNode(call.getArg(argNr))
+      or
+      // keyword argument
+      exists(Function f, string argName |
+        f = callable.getScope() and
+        f.getArgName(n) = argName and
+        result = TCfgNode(call.getArgByName(argName))
+      )
```

Comment:
This doesn't look like it will handle keyword-only parameters, since `getArgName` uses `getArg`, and `getArg` is only defined for valid positional parameters.

Reply:
Correct, this is a known limitation. Your suggestion of traversing the AST and order the arguments by location might solve this.

Reply:
Could we add that to the list of known limitations then? :wink:

Reply:
It is the only item on the list. This is the list: "A severe caveat with the approach taking in this PR is that the underlying Python library does not assign positions to keyword-only parameters (even though they clearly have a position in the definition)." It is quite possible that the list should be longer, though. For instance, there is some issue with reporting of positions depending on where arguments appear in relation to stars and double stars, and also if there is more than one star or double star. But I am not quite sure what the consequences are...

Reply:
I meant _inside the QL code_, like https://github.com/github/codeql/pull/4364/files#diff-90136ee2e7fe981452c54e6f3d2b1bf4R260-R265

Reply:
Ah, sure, I can do that :-)

Reply:
Done